src/typing_extensions.py

```diff
@@ -1236,10 +1236,91 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
             )
         else:  # 3.8
             hint = typing.get_type_hints(obj, globalns=globalns, localns=localns)
+        if sys.version_info < (3, 11):
+            _clean_optional(obj, hint, globalns, localns)
+        # types from get_type_hints might not be from a cached version
+        # In 3.8 eval_type does not handle all Optional[ForwardRef] correctly
+        # this also returns cached versions of Union and Optional
+        if sys.version_info < (3, 9):
+            hint = {
+                k: (
+                    t
+                    if get_origin(t) not in (Union, Optional)
+                    else (
+                        Optional[t.__args__[0]]
+                        if get_origin(t) == Optional
+                        else Union[t.__args__]
+                    )
+                )
+                for k, t in hint.items()
+            }
```

Comment:
Using this for other Python versions would return cached versions of `Union`. I am not 100% sure if something could go wrong here.